sdk/tests/deqp/resources/shader-library-case.js

```diff
@@ -0,0 +1,1144 @@
+    );
+
+
+};
+
+/**
+ * Execute a test case
+ * @return {boolean} True if test case passed
+ */
+var execute = function()
+{
+    /** @const @type {number} */ var quadSize = 1.0;
+    /** @const @type {Array.<number>} */
+    var s_positions = [
+        -quadSize, -quadSize, 0.0, 1.0,
+        -quadSize, +quadSize, 0.0, 1.0,
+        +quadSize, -quadSize, 0.0, 1.0,
+        +quadSize, +quadSize, 0.0, 1.0
+    ];
+
+    /** @const @type {Array.<number>} */
+    var s_indices = [
+        0, 1, 2,
+        1, 3, 2
+    ];
+
+    var wtu = WebGLTestUtils;
+    var gl = wtu.create3DContext('canvas');
+    var state = deqpTests.runner.getState();
+    var test = state.currentTest;
+    var spec = test.spec;
+
+    // Compute viewport.
+    /* TODO: original code used random number generator to compute viewport, we use whole canvas */
+    /** @const @type {number} */ var width = Math.min(canvas.width, VIEWPORT_WIDTH);
+    /** @const @type {number} */ var height = Math.min(canvas.height, VIEWPORT_HEIGHT);
+    /** @const @type {number} */ var viewportX = 0;
+    /** @const @type {number} */ var viewportY = 0;
+    /** @const @type {number} */ var numVerticesPerDraw = 4;
+    /** @const @type {boolean} */ var tessellationPresent = isTessellationPresent();
+
+    /** @type {boolean} */ var allCompilesOk = true;
+    /** @type {boolean} */ var allLinksOk = true;
+    var failReason = null;
+
+    /** @type {number} */ var vertexProgramID = -1;
+    var pipelineProgramIDs = [];
+    var programs = [];
+    var programPipeline;
+
+    // Set the name of the current test so testFailedOptions/testPassedOptions can use it.
+    setCurrentTestName(test.fullName());
+    debug('Start testcase: ' + test.fullName());
+    assertMsgOptions(gl.getError() === gl.NO_ERROR, 'Start testcase: ' + test.fullName(), false, true);
+
+    if (!test.separatePrograms)
+    {
+        var program = new deqpProgram.ShaderProgram(gl, test.programs[0].programSources);
+
+        vertexProgramID = program.getProgram();
+        pipelineProgramIDs.push(program.getProgram());
+        programs.push(program);
+
+        // Check that compile/link results are what we expect.
+
+
+        for (var i = 0; i < program.shaders.length; i++) {
+            if (!program.shaders[i].info.compileOk)
+                allCompilesOk = false;
+        }
+
+        if (!program.getProgramInfo().linkOk)
+            allLinksOk = false;
+
+
+    }
+    else
+    {
+        /* TODO: GLES 3.1: Port program pipeline code */
+    }
+
+    switch (spec.expectResult)
+    {
+        case expectResult.EXPECT_PASS:
+        case expectResult.EXPECT_VALIDATION_FAIL:
+            if (!allCompilesOk)
+                failReason = 'expected shaders to compile and link properly, but failed to compile.';
+            else if (!allLinksOk)
+                failReason = 'expected shaders to compile and link properly, but failed to link.';
+            break;
+
+        case expectResult.EXPECT_COMPILE_FAIL:
```

Comment:
It's not possible in OpenGL ES to expect compiles to fail; it's always legal for the shader compiler to defer compiles until link time. How do the tests use these expectations? I'd assume the negative tests should always use EXPECT_COMPILE_LINK_FAIL.